File: leetcode/problems/cpp/reverse-substrings-between-each-pair-of-parentheses.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "algorithm"
#include "stack"
#include "string"
using namespace std;

class Solution {
  public:
   string reverseParentheses(string s) {
      stack<int> st;
      for (int i = 0; i < s.size(); i++) {
         if (s[i] == '(')
            st.push(i);
         else if (s[i] == ')') {
            int j = st.top();
            st.pop();
            reverse(s.begin() + j, s.begin() + i);
         }
      }
      string ans = "";
      for (int i = 0; i < s.size(); i++)
         if (s[i] != '(' && s[i] != ')')
            ans += s[i];

      return ans;
   }
};

File: leetcode/problems/java/reverse-substrings-between-each-pair-of-parentheses.java
import java.util.*;

class Solution {
  public String reverseParentheses(String s) {
    ArrayDeque<String> stack = new ArrayDeque<>();
    StringBuilder sb = new StringBuilder();
    for (char c : s.toCharArray()) {
      if (c == '(') {
        stack.push(sb.toString());
        sb.setLength(0);
      } else if (c == ')') {
        sb.reverse();
        sb.insert(0, stack.pop());
      } else {
        sb.append(c);
      }
    }
    return sb.toString();
  }
}